code/datums/components/full_auto_fire.dm

```diff
@@ -0,0 +1,362 @@
+	RegisterSignal(parent, COMSIG_GUN_FIREMODE_TOGGLE, .proc/wake_up)
+	RegisterSignal(parent, COMSIG_GUN_FIREDELAY_MODIFIED, .proc/modify_firedelay)
+	RegisterSignal(parent, COMSIG_GUN_BURSTDELAY_MODIFIED, .proc/modify_burst_delay)
+	RegisterSignal(parent, COMSIG_GUN_BURSTAMOUNT_MODIFIED, .proc/modify_burst_amount)
+
+	src.autofire_shot_delay = autofire_shot_delay
+	src.burstfire_shot_delay = burstfire_shot_delay
+	src.shots_to_fire = shots_to_fire
+
+	switch(firemode)
+		if(GUN_FIREMODE_AUTOMATIC, GUN_FIREMODE_AUTOBURST)
+			var/usercli
+			if(ismob(parent_loc))
+				var/mob/user = parent_loc
+				usercli = user.client
+			wake_up(src, firemode, usercli)
+
+
+/datum/component/automatic_fire/Destroy()
+	autofire_off()
+	return ..()
+
+
+/datum/component/automatic_fire/proc/wake_up(datum/source, fire_mode, client/usercli)
+	switch(fire_mode)
+		if(GUN_FIREMODE_SEMIAUTO, GUN_FIREMODE_BURSTFIRE)
+			switch(autofire_stat)
+				if(AUTOFIRE_STAT_IDLE, AUTOFIRE_STAT_ALERT, AUTOFIRE_STAT_FIRING)
+					sleep_up()
+			return //No need for autofire on this mode.
+	
+	component_fire_mode = fire_mode
+	
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_IDLE, AUTOFIRE_STAT_ALERT)
+			return //We've updated the firemode. No need for more.
+		if(AUTOFIRE_STAT_FIRING)
+			stop_autofiring() //Let's stop shooting to avoid issues.
+			return
+	
+	autofire_stat = AUTOFIRE_STAT_IDLE
+
+	RegisterSignal(parent, list(COMSIG_PARENT_QDELETED), .proc/sleep_up)
+	RegisterSignal(parent, COMSIG_ITEM_EQUIPPED, .proc/itemgun_equipped)
+
+	if(usercli)
+		var/obj/item/weapon/gun/shoota = parent
+		if(shoota.loc == usercli.mob)
+			var/mob/shooter = usercli.mob
+			if(shooter.l_hand == parent || shooter.r_hand == parent)
+				autofire_on(usercli)
+
+
+/datum/component/automatic_fire/proc/sleep_up()
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_SLEEPING)
+			return //Already asleep
+		if(AUTOFIRE_STAT_IDLE, AUTOFIRE_STAT_ALERT, AUTOFIRE_STAT_FIRING)
+			autofire_off()
+
+	UnregisterSignal(parent, list(COMSIG_PARENT_QDELETED, COMSIG_ITEM_EQUIPPED))
+	
+	autofire_stat = AUTOFIRE_STAT_SLEEPING
+
+
+// There is a gun and there is a user wielding it. The component now waits for the mouse click.
+/datum/component/automatic_fire/proc/autofire_on(client/usercli)
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_ALERT, AUTOFIRE_STAT_FIRING)
+			return
+	autofire_stat = AUTOFIRE_STAT_ALERT
+	clicker = usercli
+	shooter = clicker.mob
+	RegisterSignal(clicker, COMSIG_CLIENT_MOUSEDOWN, .proc/on_mouse_down)
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/autofire_off)
+	RegisterSignal(shooter, COMSIG_MOB_LOGOUT, .proc/autofire_off)
+	parent.RegisterSignal(src, COMSIG_AUTOFIRE_ONMOUSEDOWN, /obj/item/weapon/gun/.proc/autofire_bypass_check)
+	parent.RegisterSignal(parent, COMSIG_AUTOFIRE_SHOT, /obj/item/weapon/gun/.proc/do_autofire)
+
+
+/datum/component/automatic_fire/proc/autofire_off(datum/source)
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_SLEEPING, AUTOFIRE_STAT_IDLE)
+			return

```

Comment:
This would need to be replaced by two `if` checks:
```
if(autofire_stat & AUTOFIRE_STAT_SLEEPING|AUTOFIRE_STAT_IDLE)
	return
if(autofire_stat & AUTOFIRE_STAT_FIRING)
	stop_autofiring()
```
Feels like a `switch` is better. You sure you want this one changed?